File: android/app/src/main/java/com/penndev/prism/data/SubscriptionParser.kt
package com.penndev.prism.data

import java.net.URI
import java.util.Base64
import org.json.JSONArray
import org.json.JSONObject

class SubscriptionException(
    val messageRes: Int,
    vararg val formatArgs: Any,
) : Exception()

object SubscriptionParser {
    fun parseContent(type: String, raw: String): List<ServerItem> {
        val text = raw.trim()
        if (text.isEmpty()) {
            throw SubscriptionException(com.penndev.prism.R.string.subscribe_error_empty)
        }
        if (text.startsWith("[")) {
            return serversFromJson(text)
        }
        return when (type.lowercase()) {
            "prism" -> parsePrism(text)
            "shadowrocket" -> parseShadowrocket(text)
            else -> throw SubscriptionException(com.penndev.prism.R.string.subscribe_error_type)
        }
    }

    fun exportJson(servers: List<ServerItem>): String {
        val arr = JSONArray()
        servers.forEach { server ->
            arr.put(
                JSONObject()
                    .put("host", server.host)
                    .put("remark", server.remark)
                    .put("username", server.username)
                    .put("password", server.password)
                    .put("protocol", server.protocol),
            )
        }
        return arr.toString(2)
    }

    fun normalizeProtocol(raw: String): String {
        val key = raw.trim().lowercase()
        return if (key in PROXY_SCHEMES) key else "socks5"
    }

    fun identity(host: String, protocol: String, username: String, password: String): String {
        val scheme = normalizeProtocol(protocol).lowercase()
        return "$scheme://$username:$password@$host"
    }

    private fun parsePrism(text: String): List<ServerItem> {
        val json = if (text.startsWith("[")) text else decodeMaybeBase64(text)
        return serversFromJson(json)
    }

    private fun parseShadowrocket(text: String): List<ServerItem> {
        val payload = decodeMaybeBase64(text)
        val seen = mutableSetOf<String>()
        val servers = payload
            .replace("\r\n", "\n")
            .split('\n')
            .mapNotNull { parseShadowrocketLine(it) }
            .filter { seen.add(it.id) }
        if (servers.isEmpty()) {
            throw SubscriptionException(com.penndev.prism.R.string.subscribe_error_no_nodes)
        }
        return servers
    }

    private fun parseShadowrocketLine(raw: String): ServerItem? {
        val uri = parseNodeUri(raw.trim()) ?: return null
        val scheme = uri.scheme?.lowercase().orEmpty()
        val protocol = when (scheme) {
            "https" -> "https"
            "socks5s" -> "socks5s"
            // Shadowrocket 的 socks5:// 是明文 SOCKS5；带 TLS 的才是 socks5s://
            "socks5", "socks" -> "socks5"
            else -> return null
        }
        val hostname = uri.host ?: return null
        if (uri.port <= 0) return null
        val host = if (hostname.contains(':')) "[$hostname]:${uri.port}" else "$hostname:${uri.port}"
        if (!HOST_PATTERN.matches(host)) return null
        val userInfo = uri.userInfo.orEmpty()
        val username = userInfo.substringBefore(':', missingDelimiterValue = userInfo)
        val password = if (userInfo.contains(':')) userInfo.substringAfter(':') else ""
        val remark = uri.fragment?.trim().orEmpty().ifBlank { host }
        return ServerItem(
            id = identity(host, protocol, username, password),
            host = host,
            remark = remark,
            username = username,
            password = password,
            protocol = protocol,
        )
    }

    private fun parseNodeUri(line: String): URI? {
        if (line.isEmpty() || !line.contains("://")) return null
        val scheme = line.substringBefore("://")
        val payload = line.substringAfter("://")
        if (scheme.isBlank() || payload.isBlank()) return runCatching { URI(line) }.getOrNull()
        val decoded = runCatching { decodeBase64Bytes(payload) }.getOrNull()
        val rebuilt = if (decoded != null) {
            "$scheme://${String(decoded, Charsets.UTF_8).trim()}"
        } else {
            line
        }
        return runCatching { URI(rebuilt) }.getOrNull()
    }

    private fun serversFromJson(json: String): List<ServerItem> {
        val arr = try {
            JSONArray(json)
        } catch (_: Exception) {
            throw SubscriptionException(com.penndev.prism.R.string.subscribe_error_json)
        }
        val servers = buildList {
            val seen = mutableSetOf<String>()
            for (i in 0 until arr.length()) {
                val obj = arr.optJSONObject(i) ?: continue
                val host = obj.optString("host").trim()
                if (host.isEmpty()) continue
                val protocol = normalizeProtocol(obj.optString("protocol", "socks5"))
                val username = obj.optString("username")
                val password = obj.optString("password")
                val id = identity(host, protocol, username, password)
                if (!seen.add(id)) continue
                add(
                    ServerItem(
                        id = id,
                        host = host,
                        remark = obj.optString("remark"),
                        username = username,
                        password = password,
                        protocol = protocol,
                    ),
                )
            }
        }
        if (servers.isEmpty()) {
            throw SubscriptionException(com.penndev.prism.R.string.subscribe_error_no_nodes)
        }
        return servers
    }

    private fun decodeMaybeBase64(text: String): String {
        return runCatching {
            String(decodeBase64Bytes(text), Charsets.UTF_8)
        }.getOrElse { text }
    }

    private fun decodeBase64Bytes(text: String): ByteArray {
        val compact = text.replace("\\s".toRegex(), "")
        val padded = compact + "=".repeat((4 - compact.length % 4) % 4)
        return try {
            Base64.getDecoder().decode(padded)
        } catch (_: Exception) {
            Base64.getUrlDecoder().decode(padded)
        }
    }
}

File: ios/Shared/SubscriptionParser.swift
import Foundation

struct SubscriptionError: LocalizedError {
    let key: String
    let intArg: Int?

    init(_ key: String, _ intArg: Int? = nil) {
        self.key = key
        self.intArg = intArg
    }

    var errorDescription: String? { key }
}

enum SubscriptionParser {
    static func parseContent(type: String, raw: String) throws -> [ServerItem] {
        let text = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if text.isEmpty {
            throw SubscriptionError("subscribe_error_empty")
        }
        if text.hasPrefix("[") {
            return try serversFromJson(text)
        }
        switch type.lowercased() {
        case "prism":
            return try parsePrism(text)
        case "shadowrocket":
            return try parseShadowrocket(text)
        default:
            throw SubscriptionError("subscribe_error_type")
        }
    }

    static func exportJson(_ servers: [ServerItem]) -> String {
        let arr: [[String: String]] = servers.map { server in
            [
                "host": server.host,
                "remark": server.remark,
                "username": server.username,
                "password": server.password,
                "protocol": server.protocolName,
            ]
        }
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys, .withoutEscapingSlashes]
        guard let data = try? encoder.encode(arr),
              let text = String(data: data, encoding: .utf8)
        else { return "[]" }
        return text
    }

    static func normalizeProtocol(_ raw: String) -> String {
        let key = raw.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        return proxySchemes.contains(key) ? key : "socks5"
    }

    static func identity(host: String, protocolName: String, username: String, password: String) -> String {
        let scheme = normalizeProtocol(protocolName).lowercased()
        return "\(scheme)://\(username):\(password)@\(host)"
    }

    private static func parsePrism(_ text: String) throws -> [ServerItem] {
        let json = text.hasPrefix("[") ? text : decodeMaybeBase64(text)
        return try serversFromJson(json)
    }

    private static func parseShadowrocket(_ text: String) throws -> [ServerItem] {
        let payload = decodeMaybeBase64(text)
        var seen = Set<String>()
        let servers = payload
            .replacingOccurrences(of: "\r\n", with: "\n")
            .split(separator: "\n", omittingEmptySubsequences: false)
            .compactMap { parseShadowrocketLine(String($0)) }
            .filter { seen.insert($0.id).inserted }
        if servers.isEmpty {
            throw SubscriptionError("subscribe_error_no_nodes")
        }
        return servers
    }

    private static func parseShadowrocketLine(_ raw: String) -> ServerItem? {
        let line = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let uri = parseNodeUri(line), let scheme = uri.scheme?.lowercased() else { return nil }
        let protocolName: String
        switch scheme {
        case "https": protocolName = "https"
        case "socks5s": protocolName = "socks5s"
        case "socks5", "socks": protocolName = "socks5"
        default: return nil
        }
        guard let hostname = uri.host, let port = uri.port, port > 0 else { return nil }
        let host = hostname.contains(":") ? "[\(hostname)]:\(port)" : "\(hostname):\(port)"
        if !hostMatches(host) { return nil }
        let username = uri.user ?? ""
        let password = uri.password ?? ""
        // URL.fragment 会留下 %E9%A6%99…；URLComponents.fragment 才是解码后的中文
        let remark = (uri.fragment ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        return ServerItem(
            id: identity(host: host, protocolName: protocolName, username: username, password: password),
            host: host,
            remark: remark.isEmpty ? host : remark,
            username: username,
            password: password,
            protocolName: protocolName,
        )
    }

    private static func parseNodeUri(_ line: String) -> URLComponents? {
        if line.isEmpty || !line.contains("://") { return nil }
        let scheme = String(line.prefix { $0 != ":" && $0 != "/" })
        guard let range = line.range(of: "://") else { return URLComponents(string: line) }
        let payload = String(line[range.upperBound...])
        if scheme.isEmpty || payload.isEmpty { return URLComponents(string: line) }
        let rebuilt: String
        if let decoded = decodeBase64Bytes(payload), let text = decodeText(decoded) {
            rebuilt = "\(scheme)://\(text.trimmingCharacters(in: .whitespacesAndNewlines))"
        } else {
            rebuilt = line
        }
        if let components = URLComponents(string: rebuilt) { return components }
        guard let url = URL(string: rebuilt) else { return nil }
        var components = URLComponents()
        components.scheme = url.scheme
        components.host = url.host
        components.port = url.port
        components.user = url.user.flatMap { $0.removingPercentEncoding } ?? url.user
        components.password = url.password.flatMap { $0.removingPercentEncoding } ?? url.password
        components.fragment = url.fragment.flatMap { $0.removingPercentEncoding } ?? url.fragment
        return components
    }

    private static func serversFromJson(_ json: String) throws -> [ServerItem] {
        guard let data = json.data(using: .utf8),
              let arr = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]]
        else { throw SubscriptionError("subscribe_error_json") }
        var seen = Set<String>()
        var servers: [ServerItem] = []
        for obj in arr {
            let host = (obj["host"] as? String ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
            if host.isEmpty { continue }
            let protocolName = normalizeProtocol(obj["protocol"] as? String ?? "socks5")
            let username = obj["username"] as? String ?? ""
            let password = obj["password"] as? String ?? ""
            let id = identity(host: host, protocolName: protocolName, username: username, password: password)
            if !seen.insert(id).inserted { continue }
            let remark = obj["remark"] as? String ?? ""
            servers.append(
                ServerItem(
                    id: id,
                    host: host,
                    remark: remark.removingPercentEncoding ?? remark,
                    username: username,
                    password: password,
                    protocolName: protocolName,
                )
            )
        }
        if servers.isEmpty {
            throw SubscriptionError("subscribe_error_no_nodes")
        }
        return servers
    }

    private static func decodeMaybeBase64(_ text: String) -> String {
        if let data = decodeBase64Bytes(text), let s = decodeText(data) {
            return s
        }
        return text
    }

    static func decodeText(_ data: Data) -> String? {
        if let s = String(data: data, encoding: .utf8) { return s }
        let gb = CFStringConvertEncodingToNSStringEncoding(
            CFStringEncoding(CFStringEncodings.GB_18030_2000.rawValue)
        )
        return String(data: data, encoding: String.Encoding(rawValue: gb))
    }

    private static func decodeBase64Bytes(_ text: String) -> Data? {
        let compact = text.replacingOccurrences(of: "\\s", with: "", options: .regularExpression)
        let pad = String(repeating: "=", count: (4 - compact.count % 4) % 4)
        let padded = compact + pad
        if let data = Data(base64Encoded: padded) { return data }
        let urlSafe = padded.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        return Data(base64Encoded: urlSafe)
    }
}
